README.md
# Vi Router

[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit)](https://github.com/pre-commit/pre-commit)
[![Go Reference](https://pkg.go.dev/badge/github.com/diontr00/vi.svg)](https://pkg.go.dev/github.com/diontr00/vi)
![ci workflow](https://github.com/diontr00/vi/actions/workflows/ci.yml/badge.svg)
[![codecov](https://codecov.io/gh/diontr00/vi/graph/badge.svg?token=bPz6VDXHae)](https://codecov.io/gh/diontr00/vi)

"This project draws inspiration from **gorilla/mux**, and I appreciate
the ease of working with the library.However, its performance
falls short of my expectations. After conducting research, my goal
is to develop a high-performance HTTP router that not only
outperforms but also retains the convenient API of mux,
enhanced with additional support for regex. I welcome any feedback
as this will be my first open source projects."

## Installation

```go
go get -u github.com/diontr00/vi
```

## TestRegex

If you want to make sure whether the regex matcher work as you expected , you can use **TestMatcher** function in the init function

```go
func init() {
    RegisterHelper("ip", `((25[0-5]|2[0-4][0-9]|1[0-9][0-9]|
    [1-9][0-9]|[0-9])\.){3}(25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])`)
    RegisterHelper("phone", `([+]?[\s0-9]+)?(\d{3}|[(]?[0-9]+[)])?([-]?[\s]?[0-9])+`)

    // Expect match
    errs := TestMatcher(true, "/user/{id:[0-9]+}/:ip/:phone?", map[vi.TestUrl]vi.TestResult{
        "/user/101/192.168.0.1/999-9999999":
        { "id" : "101" , "ip": "192.168.0.1", "phone": "999-9999999"},
        "/user/102/192.168.0.2/+48(12)504-203-260":
        { "id" : "102" , "ip": "192.168.0.2", "phone": "+48(12)504-203-260"},
        "/user/103/192.168.0.3":
        { "id" : "103" , "ip": "192.168.0.3"},
        "/user/104/192.168.0.4/555-5555-555" :
        { "id" : "104" , "ip": "192.168.0.4" ,"phone": "555-5555-555"}
 })

    for i := range errs {
        fmt.Println(errs[i])
    }

    // Expect not match
    errs := TestMatcher(false , ....)
}
```

## Simple Usage

```go
package main

import (
    "github.com/diontr00/vi"
)

func main(){
    customNotFoundHandler  := func(w http.ResponseWriter , r *http.Request) {
   w.WriteHeader(http.StatusNotFound)
   w.Write([]byte(customNotFoundMsg))
    }

    mux := vi.New(&vi.Config{Banner : false ,  NotFoundHandler: customNotFoundHandler})
    vi.RegisterHelper("ip" , `\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}`)
    mux.Get("/location/:ip", func(w http.ResponseWriter , r *http.Request){
            ip := vi.GetParam("ip")
            msg := fmt.Sprintf("You have search ip addres %s \n", ip)
            w.Write([]byte(msg))
    })

    srv :=  &http.Server{
        Addre: ":8080" ,
        Handler: vi,
    }

    done := make(chan os.Signal, 1)
 signal.Notify(done, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)

 go func() {
  if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
   log.Fatalf("listen: %s\n", err)
  }
 }()
 log.Print("Server Started")

 <-done
 log.Print("Server Stopped")

 ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 defer func() {
  // extra handling here
  cancel()
 }()

 if err := srv.Shutdown(ctx); err != nil {
  log.Fatalf("Server Shutdown Failed:%+v", err)
 }
 log.Print("Server Exited Properly")
}

```

## Serving Static Files

This receipt will serve any "userimage".png file under userfile static folder.
Note that prefix is use to append to asset path, so userimage.png eventually
become /static/userfile/userimage.png

Other properties:

- **Root:** Can be either http.Dir or embed with go embed fs
- **MaxAge:** Can be set in term of second to control the cache header
- **Next:** Skip function , will skip when query param ignore is true in this scenario
- **NotFoundFile:** if not found , it will be serve

```go
func main() {
    vi :=  vi.New()
    vi.Static("/{userimage:[0-9a-zA-Z_.]+}.png" , &StaticConfig{
        Root : http.Dir("static"),
        Prefix: "userfile",
        Index : "index.html",
        MaxAge: 2,
        Next : func(w http.ResponseWriter , r *http.Request) bool {
            return r.URL.Query().Get("ignore") == "true"
        },
        NotFoundFile : "error/notfound.html"
    } )

    http.ListenAndServe(":8080", vi)
}


```

## Matching Rule

- **Named parameter**

  - **Syntax:** :name
  - **Example:** /student/:name
  - **Explain:** Match name as word

- **Name with regex pattern**

  - **Syntax:** {name:regex-pattern}
  - **Example:** /student/{id:[0-9]+}
  - **Explain:** Match id as number

- **Helper pattern**
  - **:id** : short for **/student/{id:[0-9]+}**
  - **:name** : short for **/{name:[0-9a-zA-Z]+}**

## Benchmark

Run benchmark and test with ginkgo:

```
ginkgo
```

![](https://i.imgur.com/sxkEBvu.png)


CHANGELOG.md
## v0.4.1 (2023-12-19)

### Refactor

- ♻️ clean Static method

## v0.4.0 (2023-12-16)

### Feat

- ✨ Adding static file server functionality

### Fix

- 🐛 Fix bug associate with Static method

## v0.3.2 (2023-11-26)

### Fix

- **regex**: 🐛 fix when group use in regex pattern , submatch lead to invalid memory address access

## v0.3.1 (2023-11-25)

### Refactor

- **vi**: ♻️ make it easier to register helper pattern

## v0.3.0 (2023-11-23)

### Feat

- **router**: ✨ grouping and middleware

## v0.2.0 (2023-11-18)

### Feat

- **router**: ✨ routing Tree

## v0.1.0 (2023-11-17)

### Feat

- **regex**: ✨ regex-matcher


vi_test.go
package vi

import (
	"errors"
	"net/http"
	"net/http/httptest"
	"os"
	"strconv"

	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

var _ = DescribeTable("Test with no params", func(url, path string, expectFail bool) {
	checkSimpleResponse(url, path, expectFail)
},
	Entry("url match path", "/hello", "/hello", false),
	Entry("url not match path", "/world", "/hello", true),
)

var _ = DescribeTable("Test with params", func(url string, path string, expectParams map[matchKey]string, expectMatch bool) {
	checkResponseWithParam(url, path, expectParams, expectMatch)

},
	// Test simple param

	Entry("", "/anh", "/:name", map[matchKey]string{"name": "anh"}, true),
	Entry("", "/user/anh", "/user/:name", map[matchKey]string{"name": "anh"}, true),
	Entry("", "/user/anh/101", "/user/:name/:id", map[matchKey]string{"name": "anh", "id": "101"}, true),
	Entry("", "/video/anh/fitness", "/video/:name/:category", map[matchKey]string{"name": "anh", "category": "fitness"}, true),
	Entry("", "/video/anh/fitness", "/video/:name/:category", map[matchKey]string{"name": "anh", "category": "fitness"}, true),

	// Test param with modifier
	Entry("", "/anh", "/:name?", map[matchKey]string{"name": "anh"}, true),
	Entry("", "/user/anh", "/user/:name*", map[matchKey]string{"name": "anh"}, true),
	Entry("", "/user/anh/101", "/user/:name?/:id", map[matchKey]string{"name": "anh", "id": "101"}, true),
	Entry("", "/video/anh", "/*", nil, true),

	// Test param with regex
	Entry("", "/anh", `/{name:[\w]+}`, map[matchKey]string{"name": "anh"}, true),
	Entry("", "/user/anh", "/user/{name:[a-zA-Z]+}", map[matchKey]string{"name": "anh"}, true),
	Entry("", "/user/anh/101", "/user/:name/{id:[0-9]+}", map[matchKey]string{"name": "anh", "id": "101"}, true),

	// Other cases

	Entry("", "/anh", `/?`, nil, false),
	Entry("", "/user/anh", `/us/{name:\w+}`, nil, false),
	Entry("", "/user", "user", nil, true),
)

var _ = Describe("Test with middlewares", func() {
	var mock *handlerMock

	BeforeEach(func() {
		mock = newMock()
		mock.WrapMW("/", func(next http.HandlerFunc) http.HandlerFunc {
			return next
		})
	})

	It("Should call the register middleware", func() {

		v := New(&Config{Banner: false})
		v.Use(mock.GetMw("/"))
		mock.On("CallMock", "/").Return()
		v.GET("/", func(w http.ResponseWriter, r *http.Request) {
			return
		})

		req := httptest.NewRequest("GET", "/", http.NoBody)
		rec := httptest.NewRecorder()
		v.ServeHTTP(rec, req)
		mock.AssertCalled(GinkgoT(), "CallMock", "/")
	})

	It("Should call nested register middleware", func() {

		mock.WrapMW("/hello", func(next http.HandlerFunc) http.HandlerFunc {
			return next
		})

		mock.WrapMW("/hello/world", func(next http.HandlerFunc) http.HandlerFunc {
			return next
		})

		customNotFoundMsg := "not found handler called"
		v := New(&Config{Banner: false, NotFoundHandler: func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusNotFound)
			w.Write([]byte(customNotFoundMsg))
		}})

		v.Use(mock.GetMw("/"))

		sv := v.Group("/hello")
		ssv := sv.Group("world")
		sv.Use(mock.GetMw("/hello"))
		ssv.Use(mock.GetMw("/hello/world"))

		mock.On("CallMock", "/").Return()
		mock.On("CallMock", "/hello").Return()
		mock.On("CallMock", "/hello/world").Return()

		sv.GET("/hello/world", func(w http.ResponseWriter, r *http.Request) {
			return
		})

		ssv.GET("/hello/world/2", func(w http.ResponseWriter, r *http.Request) {
			return
		})

		req := httptest.NewRequest("GET", "/hello/world/2", http.NoBody)
		rec := httptest.NewRecorder()
		v.ServeHTTP(rec, req)
		mock.AssertCalled(GinkgoT(), "CallMock", "/")
		mock.AssertCalled(GinkgoT(), "CallMock", "/hello")
		mock.AssertCalled(GinkgoT(), "CallMock", "/hello/world")

		req = httptest.NewRequest("GET", "/notfound", http.NoBody)
		rec = httptest.NewRecorder()

		sv.ServeHTTP(rec, req)

		Expect(rec.Result().StatusCode).To(Equal(404))
		Expect(rec.Body.String()).To(Equal(customNotFoundMsg))

	})

	It("Test Coverage", func() {
		_ = New(&Config{Banner: true})
		r := httptest.NewRequest("GET", "/", http.NoBody)
		s := GetParam(r, "not-exit")
		Expect(s).To(BeZero())
	})
	It("Panic Case", func() {
		r := New(&Config{Banner: false})
		Ω(func() { r.Add("", "/", func(w http.ResponseWriter, r *http.Request) {}) }).Should(Panic())
		Ω(func() { r.Add("GET", "", func(w http.ResponseWriter, r *http.Request) {}) }).Should(Panic())

		Ω(func() { r.Add("GET", "/", nil) }).Should(Panic())
	})

})

var _ = Describe("Serving Static with simple file content and header", func() {
	maxAge := 2
	v := New(&Config{Banner: false})

	v.Static("/", &StaticConfig{Root: http.Dir("./.github/testdata/fs"), MaxAge: maxAge, NotFoundFile: "error/error.html"})

	cacheControl := "public, max-age=" + strconv.Itoa(maxAge)

	DescribeTable("", func(url string, expectStatusCode int, contentType string) {

		req := httptest.NewRequest("GET", url, http.NoBody)
		rec := httptest.NewRecorder()
		v.ServeHTTP(rec, req)

		Expect(rec.Result().StatusCode).To(Equal(expectStatusCode))
		Expect(rec.Header().Get("Content-Type")).To(Equal(contentType), "expect return correct Content-Type header")
		Expect(rec.Header().Get("Cache-Control")).To(Equal(cacheControl), "expect return correct Cache-Control header")

		var testFilePath string
		if rec.Result().StatusCode == 404 {
			testFilePath = "./.github/testdata/fs/error/error.html"
		} else {
			testFilePath = "./.github/testdata/fs" + url
		}

		file, err := os.Open(testFilePath)
		Expect(err).ToNot(HaveOccurred(), "[Setup] test file content %s should be readable : %v", testFilePath, err)

		stat, _ := file.Stat()
		var b = make([]byte, stat.Size())
		file.Read(b)
		Expect(string(b)).To(Equal(rec.Body.String()), "expect return correct file content")

	},
		Entry("When require index.html file", "/index.html", 200, "text/html"),
		Entry("When require style.css file", "/css/style.css", 200, "text/css"),
		Entry("When require index.js file", "/src/index.js", 200, "text/javascript"),
		Entry("When require notfound.js file", "/src/notfound.js", 404, "text/html"),
	)

})

var _ = Describe("Serving Static with prefix and next function", func() {
	maxAge := 2

	cacheControl := "public, max-age=" + strconv.Itoa(maxAge)
	v := New(&Config{Banner: false})

	nextFunc := func(w http.ResponseWriter, r *http.Request) bool {
		return r.URL.Query().Get("ignore") == "true"
	}

	v.Static("/", &StaticConfig{Root: http.Dir("./.github/testdata/fs"), MaxAge: maxAge, Next: nextFunc, Index: "index.html", Prefix: "sub"})

	DescribeTable("", func(url string, expectStatusCode int, contentType string, expectSkip bool) {

		prefix := "/sub"
		req := httptest.NewRequest("GET", url, http.NoBody)
		rec := httptest.NewRecorder()
		v.ServeHTTP(rec, req)

		if expectSkip {
			Expect(rec.Result().StatusCode).To(Equal(http.StatusNoContent))
			Expect(rec.Body.Available()).To(Equal(0))
			return
		}

		if rec.Result().StatusCode == 404 {
			Expect(rec.Body.String()).To(Equal("404 Not Found"), "expect default error message")
			return
		}

		Expect(rec.Result().StatusCode).To(Equal(expectStatusCode))
		Expect(rec.Header().Get("Content-Type")).To(Equal(contentType), "expect return correct Content-Type header")
		Expect(rec.Header().Get("Cache-Control")).To(Equal(cacheControl), "expect return correct Cache-Control header")

		testFilePath := "./.github/testdata/fs" + prefix + url

		file, err := os.Open(testFilePath)
		Expect(err).ToNot(HaveOccurred(), "[Setup] test file content %s should be readable : %v", testFilePath, err)

		stat, _ := file.Stat()
		var b = make([]byte, stat.Size())
		file.Read(b)
		Expect(string(b)).To(Equal(rec.Body.String()), "expect return correct file content")

	},
		Entry("When require index.html file", "/index.html", 200, "text/html", false),
		Entry("Expect skip", "/index.html?ignore=true", http.StatusNoContent, "", true),
		Entry("not found", "/notfound.js", 404, "text/plain", false),
	)

})

// EDGE CASES
type errorFS struct{}

func (e errorFS) Open(name string) (http.File, error) {
	return nil, errors.New("custom error for testing")
}

var _ = Describe("Edge Cases", func() {
	var v *vi
	var rec *httptest.ResponseRecorder
	BeforeEach(func() {
		v = New(&Config{Banner: false})
		rec = httptest.NewRecorder()

	})

	It("Should be panic if root is nill", func() {
		Ω(func() {
			v.Static("/", &StaticConfig{Root: nil})
		}).Should(Panic())
	})

	It("Should return internal error when request file cannot be open", func() {

		v.Static("/", &StaticConfig{Root: errorFS{}})
		req := httptest.NewRequest("GET", "/", http.NoBody)
		v.ServeHTTP(rec, req)
		Expect(rec.Result().StatusCode).To(Equal(http.StatusInternalServerError))
		Expect(rec.Body.String()).To(Equal("500 server internal error"))
	})
	It("Should return not found when receive method other then GET", func() {
		v.Static("/", &StaticConfig{Root: http.Dir("./.github/testdata/fs")})
		req := httptest.NewRequest("DELETE", "/index.html", http.NoBody)

		v.ServeHTTP(rec, req)

		Expect(rec.Result().StatusCode).To(Equal(http.StatusNotFound))
	})

	It("if file is dir , return index", func() {
		v.Static("/", &StaticConfig{Root: http.Dir("./.github/testdata/fs/")})
		req := httptest.NewRequest("GET", "/css", http.NoBody)

		v.ServeHTTP(rec, req)

		Expect(rec.Result().StatusCode).To(Equal(http.StatusOK))
		Expect(rec.Result().Header.Get("Content-Type")).To(Equal("text/html"))
	})

	It("if index file couldn't be open", func() {
		v.Static("/", &StaticConfig{Root: http.Dir("./.github/testdata/fs/"), Index: "notfoundindex.html"})
		req := httptest.NewRequest("GET", "/css", http.NoBody)

		Ω(func() {
			v.ServeHTTP(rec, req)
		}).Should(Panic())

	})

})


tree_test.go
package vi

import (
	"fmt"
	"strings"

	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"

	"net/http"
)

// helper to check the  where two paths diverge , so at that point more node will be create
// therefore node to be create =  len(newP) - level
func divergeAt(oldP, newP string) (level int) {
	cmap := make(map[rune]int)

	for _, rs := range oldP + newP {
		cmap[rs]++
	}

	for i, v := range oldP {
		if cmap[v] == 1 {
			return i
		}
		delete(cmap, v)
	}

	for i, v := range newP {
		if _, ok := cmap[v]; ok {
			level = i
			return
		}
	}

	return -1
}

var _ = Describe("tree: unix-test", Ordered, func() {
	var (
		nTree               *tree
		root                *treenode
		nTreeSizeInitial    int
		nTreeSizeAfterFirst int
		handler             http.HandlerFunc
		path1               = "/hello"
		path2               = "/hello2"
	)
	BeforeAll(func() {
		handler = func(w http.ResponseWriter, r *http.Request) {

			fmt.Fprint(w, "ok")
		}
		nTree = newTree()
		root = nTree.root
		Expect(nTree).To(Not(BeNil()), "tree should not be nil when initialize")
		Expect(nTree.size).To(Equal(1), "initial tree size should be 1")
		Expect(nTree.root.depth).To(Equal(1), "initial root depth should be 1")
		Expect(nTree.root.children).To(BeEmpty(), "initial children of root should be 0")
		nTreeSizeInitial = nTree.size

	})

	When("Adding node , it should be reuse , added value  and order correctly to the tree", func() {
		When("Adding initial path", func() {
			path1Len := len(trimPrefix(path1))

			mock := setupMock(handler, path1)

			It(fmt.Sprintf("Should create %d nodes when adding path : %s", path1Len, path1), func() {

				nTree.add(path1, mock.GetHandler(), []string{"/", "/path1"})
				currentNtreeSize := nTree.size
				diffLevel := divergeAt(path1, root.key)
				if diffLevel < 0 {
					// no diverge , so len(path) of nodes need to be create
					diffLevel = len(path1)
				}
				expectSize := nTreeSizeInitial + len(path1) - diffLevel

				Expect(currentNtreeSize).To(Equal(expectSize), "Tree size should be %d after create %d nodes for path : %s", expectSize, len(path1)-diffLevel, path1)
				nTreeSizeAfterFirst = currentNtreeSize
			})
			It("Added correctly", func() {
				node := root

				for i, key := range trimPrefix(path1) {
					if !node.isLeaf {
						By("Create correct amount of  node")
						Expect(node.children).To(HaveLen(1), fmt.Sprintf("Node %v should have a single children with %v \n", string(node.key), key))

						By("Only the final node will be leaf node ")
						Expect(node.isLeaf).To(BeFalse(), "node %v is not leaf node", string(node.key))
						Expect(node.path).To(BeEmpty(), "node %v is not leaf node therefore path should be empty", string(node.key))

						By("Added correct childrens")
						Expect(node.depth).To(Equal(i+1), "the depth level of node %v should be %d", string(node.key), i+1)
						children, childrenExist := node.children[nodeKey(key)]
						Expect(childrenExist).To(BeTrue(), "the children of node %v should exist", string(node.key))

						node = children
					}

				}

				hlr := node.handler
				Expect(hlr).ToNot(BeNil(), "leaf node should contain added handler")

			})
		})

		When("Adding second path", func() {
			mock := setupMock(handler, path2)

			It("Should create the different (between len(path)  and the different of two paths) nodes only", func() {
				nTree.add(path2, mock.GetHandler(), []string{"/", "/path2"})
				currentSize := nTree.size
				diffLevel := divergeAt(path1, path2)
				if diffLevel < 0 {
					diffLevel = len(path2)
				}

				expectSize := nTreeSizeAfterFirst + len(path2) - diffLevel

				Expect(currentSize).To(Equal(expectSize), "Tree size should be %d after create %d nodes for path : %s", expectSize, len(path2)-diffLevel, path2)
			})

			It("Should added correctly", func() {
				node := root

				diffLevel := divergeAt(path1, path2)

				for i, key := range trimPrefix(path2) {
					if !node.isLeaf {
						By("Reuse node")
						if i == diffLevel {
							Expect(node.children).To(HaveLen(2), fmt.Sprintf("Node %d of key %s should be reuse, and therfore another children should be added", diffLevel, string(path2[diffLevel-1])))
						} else {

							Expect(node.children).To(HaveLen(1), "other newly added node will  have children of 1")
						}

						By("Only the final node will be leaf node ")
						Expect(node.isLeaf).To(BeFalse(), "node %v is not leaf node", string(node.key))
						Expect(node.path).To(BeEmpty(), "node %v is not leaf node therefore path should be empty", string(node.key))

						By("Added correct childrens")
						Expect(node.depth).To(Equal(i+1), "the depth level of node %v should be %d", string(node.key), i+1)
						children, childrenExist := node.children[nodeKey(key)]
						Expect(childrenExist).To(BeTrue(), "the children of node %v should exist", string(node.key))

						node = children

					}

				}

				hlr := node.handler
				Expect(hlr).ToNot(BeNil(), "leaf node should contain added handler")
			})
		})

	})

	When("Finding node in the tree", func() {
		It("Should return matches node", func() {
			By("Finding nodes for the first path")
			nodes := nTree.find(path1)
			Expect(nodes).ToNot(BeNil(), fmt.Sprintf("nodes for %s should not be nil", path1))
			Expect(nodes).To(HaveLen(1), "return matchexact path hello")
			Expect(string(nodes[0].key)).To(Equal("o"), "last element of path1 is o")
			By("Finding nodes for the second path")
			nodes = nTree.find(path2)
			Expect(nodes).ToNot(BeNil(), fmt.Sprintf("nodes for %s should not be nil", path2))
			Expect(nodes).To(HaveLen(1), "expect path hello2 to match only handler for hello2 not hello")
			Expect(string(nodes[0].key)).To(Equal("2"))
			By("Finding nodes for invalid path")
			nodes = nTree.find("invalid")
			Expect(nodes).To(BeEmpty(), "should not return any nodes for invalid path")

		})

	})

})

func trimPrefix(path string) string {
	return strings.Trim(path, "/")
}


internal/color/color.go
package color

import (
	"bytes"
	"fmt"
	"github.com/mattn/go-isatty"
	"os"
)

const (
	GreenFg  = "32"
	BlueFg   = "34"
	RedFg    = "31"
	BoldText = "1"
)

var (
	green = makecolor(GreenFg)
	blue  = makecolor(BlueFg)
	red   = makecolor(RedFg)
	bold  = makecolor(BoldText)

	noColor = !isatty.IsTerminal(os.Stdout.Fd()) && !isatty.IsCygwinTerminal(os.Stdout.Fd())
)

type colorize func(interface{}) string

func makecolor(n string) colorize {
	return func(msg interface{}) string {
		b := new(bytes.Buffer)
		b.WriteString("\x1b[")
		b.WriteString(n)
		b.WriteString("m")
		return fmt.Sprintf("%s%v\x1b[0m", b.String(), msg)
	}
}

func Green(format string, args ...any) string {
	m := makeMsg(format, args...)
	if noColor {
		return m
	}
	return green(m)
}

func Blue(format string, args ...any) string {
	m := makeMsg(format, args...)
	if noColor {
		return m
	}
	return blue(m)
}

func Red(format string, args ...any) string {
	m := makeMsg(format, args...)

	if noColor {
		return m
	}
	return red(m)
}

func Bold(format string, args ...any) string {
	m := makeMsg(format, args...)
	return bold(m)
}

func makeMsg(format string, args ...any) (m string) {
	if len(args) == 0 {
		m = format
	} else {
		m = fmt.Sprintf(format, args...)
	}
	return
}


vi.go
package vi

import (
	"bufio"
	"context"
	"errors"
	"fmt"
	"github.com/diontr00/vi/internal/color"
	"github.com/diontr00/vi/internal/utils"
	"io/fs"
	"log"
	"net/http"
	ospath "path"
	"strconv"
	"strings"
)

// type of the context key
type ctxKey struct{}

// use to set the key of a context
var contextKey = ctxKey{}

// Configuration for vi instance
type Config struct {
	// When set to false , this will turn off the banner
	Banner bool
	// Set the custom not found error handler , if not set the default not fault will be use
	NotFoundHandler http.HandlerFunc
}

// Static defines configuration options when defining static route
type StaticConfig struct {
	// Root is a filesystem  that provides access to a
	// collection of files and directories , use http.Dir("folder name") or embed FS  with http.FS()
	// Required
	Root http.FileSystem
	// Defines prefix that will be add to be when reading a file from FileSystem
	// Use only when using go embed FS for Root
	// Optional  default to ""
	Prefix string
	// Name of the index file for serving
	// Optional default to index.html
	Index string
	// The value for the cache-control HTTP-Header when response , its define in term of second , default value to 0
	// Optional default to 0
	MaxAge int
	//  Next defines a function  that allow to skip a scenario when it return true
	// Optional default to nil
	Next func(w http.ResponseWriter, r *http.Request) bool
	// File to return if path is not found. Useful for SPA's
	// Optional default to 404 not found
	NotFoundFile string
}

type middleware func(next http.HandlerFunc) http.HandlerFunc

type vi struct {
	// hold prefix that relevant for particular vi instance
	prefixes []string
	// routing tree
	trees map[string]*tree
	// map between prefix and middleware
	middlewares map[string][]middleware
	// not found error handler
	notfoundhandler http.HandlerFunc
}

// Return new vi
func New(config *Config) *vi {
	v := new(vi)
	v.prefixes = []string{"/"}
	v.middlewares = map[string][]middleware{"/": {}}
	v.trees = make(map[string]*tree)

	if config != nil && config.Banner {
		fmt.Println(color.Green(banner, color.Blue(Version), color.Red(website)))
	}
	if config.NotFoundHandler != nil {
		v.notfoundhandler = config.NotFoundHandler
	} else {
		v.notfoundhandler = func(w http.ResponseWriter, r *http.Request) {
			http.NotFound(w, r)
		}
	}
	return v
}

// HTTP get routing along "pattern"
func (v *vi) GET(path string, handler http.HandlerFunc) {
	v.Add("GET", path, handler)
}

// HTTP post routing along "pattern"
func (v *vi) POST(path string, handler http.HandlerFunc) {
	v.Add("POST", path, handler)
}

// HTTP put routing along "pattern"
func (v *vi) PUT(path string, handler http.HandlerFunc) {
	v.Add("PUT", path, handler)
}

// HTTP delete routing along "pattern"
func (v *vi) DELETE(path string, handler http.HandlerFunc) {
	v.Add("DELETE", path, handler)
}

// HTTP path routin  along "pattern"
func (v *vi) PATCH(path string, handler http.HandlerFunc) {
	v.Add("PATCH", path, handler)
}

// register new  HTTP verb routing along pattern
func (v *vi) Add(method, path string, handler http.HandlerFunc) {
	if method == "" {
		panic(color.Red("method must not be empty"))
	}

	if len(path) < 1 {
		panic(color.Red("path must not be empty"))
	}
	if string(path[0]) != "/" {
		path = "/" + path
	}
	if handler == nil {
		panic(color.Red("handler must not be nil"))
	}

	if v.trees == nil {
		v.trees = make(map[string]*tree)
	}

	tree := v.trees[method]
	if tree == nil {
		tree = newTree()
		v.trees[method] = tree
	}

	tree.add(path, handler, v.prefixes)
}

func (v *vi) registerStatic(path string, cfg *StaticConfig) {
	cacheControl := "public, max-age=" + strconv.Itoa(cfg.MaxAge)

	handler := func(w http.ResponseWriter, r *http.Request) {
		if cfg.Next != nil && cfg.Next(w, r) {
			w.WriteHeader(http.StatusNoContent)
			return
		}

		// path to search for static files
		var searchp string = ospath.Clean(r.URL.Path)

		if cfg.Prefix != "" {
			searchp = cfg.Prefix + searchp
		}
		if len(searchp) > 1 {
			searchp = strings.TrimSuffix(searchp, "/")
		}

		if cfg.MaxAge > 0 {
			w.Header().Set("Cache-Control", cacheControl)
		}

		file, err := cfg.Root.Open(searchp)
		if err != nil {
			if errors.Is(err, fs.ErrNotExist) {
				w.WriteHeader(http.StatusNotFound)
				w.Header().Set("Content-Type", "text/plain")
				if cfg.NotFoundFile == "" {
					w.Write([]byte("404 Not Found"))
					return
				}

				nffile, err := cfg.Root.Open(cfg.NotFoundFile)
				if err != nil {
					w.Write([]byte("404 Not Found"))
					log.Print(color.Red("[Warning] , not found file couldn't be open : %v \n", err))

					return
				}

				defer nffile.Close()
				w.Header().Set("Content-Type", utils.GetMIME(utils.GetFileExtension(cfg.NotFoundFile)))
				bufio.NewReader(nffile).WriteTo(w)
				return
			} else {
				w.WriteHeader(http.StatusInternalServerError)
				w.Header().Set("Content-Type", "text/plain")
				w.Write([]byte("500 server internal error"))
				log.Print(color.Red("[Error] , could not open file %s for reading : %v \n", color.Bold(searchp), err))
				return
			}
		}

		defer file.Close()
		stat, _ := file.Stat()

		// Serve index if path is directory
		if stat.IsDir() {
			index, err := cfg.Root.Open(cfg.Index)
			if err != nil {
				log.Panicf(color.Red("Index file couldn't be open : %v \n", err))
			}

			defer index.Close()
			file = index
			searchp = cfg.Index
		}

		mimeType := utils.GetFileExtension(searchp)
		w.Header().Set("Content-Type", utils.GetMIME(mimeType))
		if cfg.MaxAge > 0 {
			w.Header().Set("Cache-Control", cacheControl)
		}

		bufio.NewReader(file).WriteTo(w)
	}
	v.Add(http.MethodGet, path, handler)
}

// Static will create a file server serving the static file
// if path present the path pattern
func (v *vi) Static(path string, config *StaticConfig) {
	var cfg = &StaticConfig{
		Index:        "/index.html",
		MaxAge:       0,
		Next:         nil,
		NotFoundFile: "",
		Root:         nil,
		Prefix:       "",
	}

	if config != nil {
		cfg = config
		if config.Index == "" {
			cfg.Index = "index.html"
		}
		if !strings.HasPrefix(cfg.Index, "/") {
			cfg.Index = "/" + cfg.Index
		}
		if cfg.NotFoundFile != "" && !strings.HasPrefix(cfg.NotFoundFile, "/") {
			cfg.NotFoundFile = "/" + cfg.NotFoundFile
		}
	}

	if cfg.Root == nil {
		panic("Http file server root cannot be nil")
	}

	if cfg.Prefix != "" && !strings.HasPrefix(cfg.Prefix, "/") {
		cfg.Prefix = "/" + cfg.Prefix
	}
	v.registerStatic(path, cfg)
}

// use to group route under prefix
func (v *vi) Group(prefix string) *vi {
	if string(prefix[0]) != "/" {
		prefix = "/" + prefix
	}
	prefixes := v.prefixes

	if _, ok := v.middlewares[prefix]; !ok {
		v.middlewares[prefix] = make([]middleware, 0)
		prefixes = append(prefixes, prefix)
	}

	return &vi{
		prefixes:        prefixes,
		trees:           v.trees,
		middlewares:     v.middlewares,
		notfoundhandler: v.notfoundhandler,
	}
}

// use to register middlewares
func (v *vi) Use(middlewares ...middleware) {
	// get the last prefix added
	prefix := v.prefixes[len(v.prefixes)-1]

	if len(middlewares) > 0 {
		v.middlewares[prefix] = append(v.middlewares[prefix], middlewares...)
	}
}

// chain all middlewares associate with prefixes
func (v *vi) chain(w http.ResponseWriter, r *http.Request, handler http.HandlerFunc, prefixes []string) {
	var allMiddleware []middleware
	for _, p := range prefixes {
		allMiddleware = append(allMiddleware, v.middlewares[p]...)
	}

	for i := len(allMiddleware) - 1; i >= 0; i-- {
		handler = allMiddleware[i](handler)
	}

	handler(w, r)
}

// Get the matched  param that store inside request context
func GetParam(r *http.Request, key string) (paramValue string) {
	values, ok := r.Context().Value(contextKey).(matchParams)
	if ok {
		paramValue, ok = values[matchKey(key)]
	}

	if !ok {
		paramValue = ""
	}

	return paramValue
}

func (v *vi) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	rqUrl := r.URL.Path
	if _, ok := v.trees[r.Method]; !ok {
		v.trees[r.Method] = newTree()
	}
	nodes := v.trees[r.Method].find(rqUrl)
	for i := range nodes {
		handler := nodes[i].handler
		if handler != nil {
			if nodes[i].path == rqUrl {
				v.chain(w, r, handler, nodes[i].prefixes)
				return
			}
		}
	}

	if nodes == nil {
		// match against any regex match
		nodes := v.trees[r.Method].find("/")

		for i := range nodes {
			handler := nodes[i].handler

			if handler != nil {
				isMatch, matchParams := match(rqUrl, nodes[i].path)
				if isMatch {
					ctx := context.WithValue(r.Context(), contextKey, matchParams)
					r = r.WithContext(ctx)
					v.chain(w, r, handler, nodes[i].prefixes)
					return
				}
			}
		}
	}

	v.notfoundhandler(w, r)
}
